cartographer/io/serialization_format_migration_test.cc

```diff
@@ -0,0 +1,121 @@
+#include "gtest/gtest.h"
+
+namespace cartographer {
+namespace io {
+namespace {
+
+using ::google::protobuf::TextFormat;
+using ::testing::Eq;
+using ::testing::SizeIs;
+
+class MigrationTest : public ::testing::Test {
+ protected:
+  void SetUp() {
+    writer_.reset(new ForwardingProtoStreamWriter(
+        [this](const google::protobuf::Message* proto) -> bool {
+          std::string msg_string;
+          TextFormat::PrintToString(*proto, &msg_string);
+          this->output_messages_.push_back(msg_string);
+          return true;
+        }));
+
+    mapping::proto::PoseGraph pose_graph;
+    mapping::proto::AllTrajectoryBuilderOptions all_options;
+    mapping::proto::LegacySerializedData submap;
+    submap.mutable_submap()->mutable_submap_id()->set_trajectory_id(0);
+    mapping::proto::LegacySerializedData node;
+    node.mutable_node()->mutable_node_id()->set_trajectory_id(0);
+    mapping::proto::LegacySerializedData imu_data;
+    imu_data.mutable_imu_data()->set_trajectory_id(0);
+    mapping::proto::LegacySerializedData odometry_data;
+    odometry_data.mutable_odometry_data()->set_trajectory_id(0);
+    mapping::proto::LegacySerializedData fixed_frame_pose;
+    fixed_frame_pose.mutable_fixed_frame_pose_data()->set_trajectory_id(0);
+    mapping::proto::LegacySerializedData trajectory_data;
+    trajectory_data.mutable_trajectory_data()->set_trajectory_id(0);
+    mapping::proto::LegacySerializedData landmark_data;
+    landmark_data.mutable_landmark_data()->set_trajectory_id(0);
+
+    reader_.AddProto(pose_graph);
+    reader_.AddProto(all_options);
+    reader_.AddProto(submap);
+    reader_.AddProto(node);
+    reader_.AddProto(imu_data);
+    reader_.AddProto(odometry_data);
+    reader_.AddProto(fixed_frame_pose);
+    reader_.AddProto(trajectory_data);
+    reader_.AddProto(landmark_data);
+  }
+
+  InMemoryProtoStreamReader reader_;
+  std::unique_ptr<ForwardingProtoStreamWriter> writer_;
+  std::vector<std::string> output_messages_;
+
+  static constexpr int kNumOriginalMessages = 9;
+};
+
+TEST_F(MigrationTest, MigrationAddsHeaderAsFirstMessage) {
+  MigrateStreamFormatToVersion1(&reader_, writer_.get());
+  // We expect one message more than the original number of messages, because of
+  // the added header.
+  EXPECT_THAT(output_messages_, SizeIs(kNumOriginalMessages + 1));
+
+  mapping::proto::SerializationHeader header;
+  EXPECT_THAT(TextFormat::ParseFromString(output_messages_[0], &header),
+              Eq(true));
+  EXPECT_THAT(header.format_version(), Eq(1));
+}
+
+TEST_F(MigrationTest, SerializedDataOrderIsCorrect) {
+  MigrateStreamFormatToVersion1(&reader_, writer_.get());
+  EXPECT_THAT(output_messages_, SizeIs(kNumOriginalMessages + 1));
+
+  std::vector<mapping::proto::SerializedData> serialized(
+      output_messages_.size() - 1);
+  for (size_t i = 1; i < output_messages_.size(); ++i) {
+    EXPECT_THAT(
+        TextFormat::ParseFromString(output_messages_[i], &serialized[i - 1]),
```

Comment:
Just as a note, if the messages are identical, you can compare them with google::protobuf::util::MessageDifferencer.